### ISSUE TITLE
Is dlib possible to load the necessary images for each epoch?

### ISSUE THREAD

**Issue by freddi-kit** [davisking/dlib#1321]
## Expected Behavor
I want to train many images by loss mmod. I want model to learn negative situation, so I have listed many images in xml.

## Current Behavior
When dlib starts training, dlib loads all images listed in xml, so training process is killed by memory flow.

## Steps to Reproduce
Is dlib possible to load the necessary images for each epoch instead of loading all the images in memory before learning the deep learning model?

* **Version**: dlib-19.10
* **Where did you get dlib**: dlib.net
* **Platform**: Ubuntu 16.04.3 LTS 16bit
* **Compiler**: gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.9) 

Thanks!

**Comment by davisking** [comment]
Yes, of course you can do this.  Write code that only loads some images at at time rather than all of them.  "dlib" doesn't load anything.  It's the user that does the loading.  There are example program that show you how to generally use dlib and some of them load data.  If you want to load data in a different way it's on you to write a new program that does that.

**Comment by LuWei6896** [comment]
> Thank you! I'll try writing code!
 I am doing this work again,could you share your code,with my appricate!

**Comment by LuWei6896** [comment]
I am doing this work again,could you share your code,with my appricate!

**Comment by arrufat** [comment]
If you look inside of dlib's code, you'll see that most of what you want is already there.
If your detection dataset is too big for it to fit in memory, you can use `load_image_dataset_metadata`.
This is what [`load_image_dataset`](https://github.com/davisking/dlib/blob/6b581d91f6b9b847a8163420630ef947e7cc88db/dlib/data_io/load_image_dataset.h#L100-L189) does, and then it loads the images and the bounding boxes into memory.
Just modify that to return  a `std::vector` of paths instead of images.

**Comment by freddi-kit** [comment]
@LuWei6896 sorry, because I left previous company so I did not have code.

**Comment by freddi-kit** [comment]
And `load_image_dataset_metadata` looks great. I want to show it to my old self.

**Comment by LuWei6896** [comment]
@freddi-kit yes! show me the code. with my appricate!

**Comment by LuWei6896** [comment]
@freddi-kit  I see,thanks!

**Comment by LuWei6896** [comment]
@freddi-kit @arrufat could you tell how to load pretrain modle by dlib,i wang to fine-tune it !
